library/core/src/hint.rs

```diff
@@ -51,18 +51,47 @@ pub const unsafe fn unreachable_unchecked() -> ! {
     unsafe { intrinsics::unreachable() }
 }
 
-/// Emits a machine instruction hinting to the processor that it is running in busy-wait
-/// spin-loop ("spin lock").
+/// Emits a machine instruction to signal the processor that it is running in
+/// busy-wait spin-loop ("spin lock").
+///
+/// Upon receiving spin-loop signal the processor can optimize its behavior by,
+/// for example, saving power or switching hyper-threads.
+///
+/// This function is different from [`std::thread::yield_now`] which directly
+/// yields to the system's scheduler, whereas `spin_loop` does not interact
+/// with the operating system.
+///
+/// A common use case for `spin_loop` is implementing bounded optimistic
+/// spinning in a CAS loop in synchronization primitives. To avoid problems
+/// like priority inversion, it is strongly recommended that the spin loop is
+/// terminated after a finite amount of iterations and an appropriate blocking
+/// syscall is made.
+///
+/// **Note**: On platforms that do not support receiving spin-loop hints this
+/// function does not do anything at all.
 ///
-/// For a discussion of different locking strategies and their trade-offs, see
-/// [`core::sync::atomic::spin_loop_hint`].
+/// # Examples
 ///
-/// **Note**: On platforms that do not support receiving spin-loop hints this function does not
-/// do anything at all.
+/// ```
+/// use std::hint;
+/// use std::sync::atomic::{AtomicBool, Ordering};
+///
+/// let dead = AtomicBool::new(false);
```

Comment:
Let's pick a different name for this, to avoid the _death_ connection. Perhaps a tiny atomic lock like [this one](https://doc.rust-lang.org/std/sync/atomic/#examples)?

```rust
use std::sync::Arc;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::thread;

fn main() {
    // Say we have a spinlock, which can be acquired so long as its value is `0`
    // In this example, the lock is already acquired
    let spinlock = Arc::new(AtomicUsize::new(1));

    // Spawn off a thread that will release the lock at some point
    let spinlock_clone = spinlock.clone();
    let thread = thread::spawn(move|| {
        // Do some work, knowing we have the lock
        spinlock_clone.store(0, Ordering::SeqCst);
    });

    // Wait for the other thread to release the lock
    // We do this in a loop, using `hint::spin_loop` to signal that we're waiting
    while spinlock.load(Ordering::SeqCst) != 0 {
        // Signal to the processor that we're in a busy-wait loop
        hint::spin_loop();
    }
}
```

Reply:
I do see that this is a valid and easy to understand example but maybe we should provide an example with a better use case? Just by looking at this I may not know where it is best used, maybe we should have a real-life use case example here? What do you think?

I used *death* here because it everyone can understand and relate since people have 100% chance of dying. If you don't want *death*, how above *live*?

Reply:
No problem! I think "live" is ok, but would definitely like to avoid "death".